app/controllers/api/migrations.php

```diff
@@ -290,6 +300,136 @@
+        $isAPIKey = Auth::isAppUser(Authorization::getRoles());
+        $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::getRoles());
+
+        // Check if migration/import is already in progress!
+        // if for some reason the worker crashes, the stage will always be `init`, what do we do?
+        $isInProgress = Authorization::skip(function () use ($dbForProject, $resourceId) {
+            $exists = $dbForProject->findOne(
+                'migrations',
+                [
+                    Query::notEqual('stage', 'finished'),
+                    Query::equal('resourceId', [$resourceId]),
+                ]
+            );
+
+            return !$exists->isEmpty();
+        });
+
+        if ($isInProgress || (!$isAPIKey && !$isPrivilegedUser)) {
+            throw new Exception(Exception::MIGRATION_IN_PROGRESS, 'An import is already in progress for this collection.');
+        }
+
+        $bucket = Authorization::skip(fn () => $dbForProject->getDocument('buckets', $bucketId));
+
+        if ($bucket->isEmpty() || (!$isAPIKey && !$isPrivilegedUser)) {
+            throw new Exception(Exception::STORAGE_BUCKET_NOT_FOUND);
+        }
+
+        $file = Authorization::skip(fn () => $dbForProject->getDocument('bucket_' . $bucket->getInternalId(), $fileId));
+        if ($file->isEmpty()) {
+            throw new Exception(Exception::STORAGE_FILE_NOT_FOUND);
+        }
+
+        $path = $file->getAttribute('path', '');
+        if (!$deviceForFiles->exists($path)) {
+            throw new Exception(Exception::STORAGE_FILE_NOT_FOUND, 'File not found in ' . $path);
+        }
+
+        // read file content.
+        $source = $deviceForFiles->read($path);
+
+        // decrypt
+        if (!empty($file->getAttribute('openSSLCipher'))) {
+            $source = OpenSSL::decrypt(
+                $source,
+                $file->getAttribute('openSSLCipher'),
+                System::getEnv('_APP_OPENSSL_KEY_V' . $file->getAttribute('openSSLVersion')),
+                0,
+                \hex2bin($file->getAttribute('openSSLIV')),
+                \hex2bin($file->getAttribute('openSSLTag'))
+            );
+        }
+
+        // decompress
+        switch ($file->getAttribute('algorithm', Compression::NONE)) {
+            case Compression::ZSTD:
+                $compressor = new Zstd();
+                $source = $compressor->decompress($source);
+                break;
+            case Compression::GZIP:
+                $compressor = new GZIP();
+                $source = $compressor->decompress($source);
+                break;
+        }
+
+        // copy to temporary folder
+        $migrationId = ID::unique();
+        $path = $deviceForCsvImports->getRoot() . '/' . $migrationId . '_' . $fileId . '.csv';
+        $deviceForCsvImports->write($path, $source, 'text/csv');
+        $fileSize = $deviceForCsvImports->getFileSize($path);
```

Comment:
There's a helper method like Device::transfer you can use to move files between `Devices`, we can use it here once not reading into memory

Reply:
Answered above. This is needed due to decrypted/uncompressed file after a fetch from bucket.